### ISSUE TITLE
Stepheight issues with touching ground in liquid

### ISSUE THREAD

**Issue by CasimirKaPazi** [minetest/minetest#7742]
Possibly a bug/unintended behavior. In all minetest versions. I am using current dev.

In my subgame Voxelgarden I set the stepheight to 1.1. So when walking up against a full node the player steps onto it. However when swimming in liquid it's a problem. When the player swims against a node on the shore they don't step up, *until* they sank and touch ground, then suddenly the step onto the node.

In game play this is a very odd behavior. So maybe an check for the player being in liquid should be added.

This would also benefit games with smaller step height. As in MTG one always has to "jump" out of water. With a check for liquid the player would step onto the shore if they are high enough for the step height to take effect.

**Comment by SmallJoker** [comment]
Can probably be solved/worked around by the autojump feature (#7228). If you're interested in debugging this on your own, also look at #7688. I also encountered some unwanted behaviour while swimming and jumping in `localplayer.cpp` and the collision engine around it.

**Comment by bendeutsch** [comment]
I'm guessing autojump (#7228) will *not* help, since as currently written, it only tries to jump if you're standing on solid ground. Probably a better solution would be to change the "stepping" code to fire if you're on solid ground *or* swimming.

**Comment by paramat** [comment]
I don't see this as a bug, although it results in odd behaviour in the case of a 1.1 stepheight. More a request to change behaviour to have stepheight apply when you are swimming.
If the result is not having to jump out of the water with the normal stepheight this is a change in movement behaviour, so needs to be approached carefully, someone may be relying on the current behaviour.

**Comment by CasimirKaPazi** [comment]
> Edit: the step height code, as in the current master, makes a difference between "on ground" and "not" around here: https://github.com/minetest/minetest/blob/master/src/localplayer.cpp#L287

Simply adding "or in liquid" does the trick.

    float player_stepheight = (m_cao == nullptr) ? 0.0f :
    	(touching_ground || in_liquid ? m_cao->getStepHeight() : (0.2f * BS));

As far as I tested it feels much more natural.

**Comment by CasimirKaPazi** [comment]
> someone may be relying on the current behaviour.

No. There is no way to "use" it. It's just annoying and it's a bug. I tested it for several settings of step height and it works smoothly.

**Comment by paramat** [comment]
Sneak jump and sneak ladder were 'bugs' too initially.
However i agree with considering this request, carefully.
First though try the new autojump feature as it may be better than altering stepheight.

**Comment by jastevenson303** [comment]
I use it by sprinting out of the water only, but I will admit that any
other time I'm not using sneak_jump mod, it's very annoying not being able
to jump out of the water.  Sorry for splitting hairs.  :)

**Comment by CasimirKaPazi** [comment]
> First though try the new autojump feature as it may be better than altering stepheight.

This has nothing to do with this issue discussed here. You still have to press jump to get out of water.

One a side note; I also prefer stepheight over autojump. To jump all the time during the game is a strange thing to do, to automate it (instead of using an existing alternative) doesn't make it better.

**Comment by jastevenson303** [comment]
>To jump all the time during the game is a strange thing to do...

Strange to you, but a personal preference to myself.  It's the opposite feeling for me, I think autojump is odd.  I do agree adjusting stepheight is better than using autojump.

Sorry for the aside, the underlying issue remains true: 

> When the player swims against a node on the shore they don't step up, until they sank and touch ground, then suddenly the step onto the node.

>In game play this is a very odd behavior. So maybe an check for the player being in liquid should be added.

I'd like to add also that in regular MT you cannot easily get out of the water at all, if there isn't an adjacent node.  A node only one node up can't be jumped onto.

**Comment by paramat** [comment]
> This has nothing to do with this issue discussed here. You still have to press jump to get out of water.

Correct, my mistake.
